File: day01/secret_entrance.cpp
#include <stdlib.h>
#include <iostream>
#include <fstream>
#include <string>

using namespace std;

const int mod(const int a, const int b) {
    return (a % b + b) % b;
}

int main(int argc, char *argv[]) {
    const string fname = argv[1];
    int part1 = 0;
    int part2 = 0;
    int pos1 = 50;
    int pos2 = 50;

    ifstream infile(fname);
    if (!infile) {
        perror(fname.c_str());
        exit(1);
    }

    string line;
    while (!infile.eof()) {
        getline(infile, line);
        if (line.size() < 2)
            continue;
        const int dir = line[0] == 'R' ? 1 : -1;
        const int cnt = stoi(line.substr(1)) * dir;
        pos1 = (pos1 + cnt) % 100;
        if (pos1 == 0)
            part1++;

        const bool on_zero = (pos2 == 0);
        pos2 += cnt;

        if (pos2 <= 0 && !on_zero)
            part2++;

        part2 += abs(pos2) / 100;
        pos2 = mod(pos2, 100);
    }

    cout << "Part 1: " << part1 << endl;
    cout << "Part 2: " << part2 << endl;
}

File: day01/secret_entrance.py
from sys import argv

part1 = 0
part2 = 0
pos = 50
pos2 = 50
with open(argv[1]) as f:
    for line in f:
        direction = 1 if line[0] == 'R' else -1
        cnt = int(line[1:])
        pos = (pos + (cnt * direction)) % 100
        if pos == 0:
            part1 += 1

        for _ in range(cnt):
            pos2 = (pos2 + direction) % 100
            if pos2 == 0:
                part2 += 1

print('Part 1:', part1)
print('Part 2:', part2)
